stackslib/src/chainstate/nakamoto/mod.rs

```diff
@@ -1806,6 +1813,261 @@ impl NakamotoChainState {
+
+        let signers_list: Vec<_> = reward_set
+            .signers
+            .ok_or(ChainstateError::PoxNoRewardCycle)?
+            .iter()
+            .map(|signer| {
+                Value::Tuple(
+                    TupleData::from_data(vec![
+                        (
+                            "signer".into(),
+                            Value::Principal(PrincipalData::from(signer.signing_address)),
+                        ),
+                        ("num-slots".into(), Value::UInt(signer.slots.into())),
+                    ])
+                    .expect(
+                        "BUG: Failed to construct `{ signer: principal, num-slots: u64 }` tuple",
+                    ),
+                )
+            })
+            .collect();
+
+        if signers_list.len() > SIGNERS_MAX_LIST_SIZE {
+            panic!(
+                "FATAL: signers list returned by reward set calculations longer than maximum ({} > {})",
+                signers_list.len(),
+                SIGNERS_MAX_LIST_SIZE,
+            );
+        }
+
+        let args = [
+            SymbolicExpression::atom_value(Value::cons_list_unsanitized(signers_list).expect(
+                "BUG: Failed to construct `(list 4000 { signer: principal, num-slots: u64 })` list",
+            )),
+            SymbolicExpression::atom_value(Value::UInt(reward_cycle.into())),
+        ];
+
+        let (value, _, events, _) = clarity
+            .with_abort_callback(
+                |vm_env| {
+                    vm_env.execute_in_env(sender_addr.clone(), None, None, |env| {
+                        env.execute_contract_allow_private(
+                            &signers_contract,
+                            "stackerdb-set-signer-slots",
+                            &args,
+                            false,
+                        )
+                    })
+                },
+                |_, _| false,
+            )
+            .expect("FATAL: failed to update signer stackerdb");
+
+        if let Value::Response(ref data) = value {
+            if !data.committed {
+                error!(
+                    "Error while updating .signers contract";
+                    "reward_cycle" => reward_cycle,
+                    "cc_response" => %value,
+                );
+                panic!();
+            }
+        }
+
+        Ok(events)
+    }
+
+    pub fn check_and_handle_prepare_phase_start(
+        clarity_tx: &mut ClarityTx,
+        first_block_height: u64,
+        pox_constants: &PoxConstants,
+        burn_tip_height: u64,
+    ) -> Result<Vec<StacksTransactionEvent>, ChainstateError> {
+        let current_epoch = clarity_tx.get_epoch();
+        if current_epoch < StacksEpochId::Epoch25 {
+            // before Epoch-2.5, no need for special handling
+            return Ok(vec![]);
+        }
+        // now, determine if we are in a prepare phase, and we are the first
```

Comment:
Just thinking a little bit ahead here -- would it be acceptable to relax this constraint to "in or _after_ this last prepare phase?"  Asking because in the (unlikely but possible) event that the prepare phase is devoid of blocks, I think we'd want the signer set calculation to still go through during the subsequent reward phase.  Perhaps the code here could check the last-processed reward cycle against the current reward cycle, and proceed if either we're in the prepare phase for the next cycle or we're in a subsequent cycle altogether?

Reply:
Yeah, I think so, but I'll start an issue for it -- I would want to separate it from the upstream PR here.